File: Frontend/Solver/src/dev/vale/solver/SimpleSolverState.scala
package dev.vale.solver

import dev.vale.{Err, Ok, Result, vassert, vassertSome, vcurious, vfail}
import dev.vale.Err

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer


object SimpleSolverState {
  def apply[Rule, Rune, Conclusion](): SimpleSolverState[Rule, Rune, Conclusion] = {
    SimpleSolverState[Rule, Rune, Conclusion](
      Vector(),
      Map[Rune, Int](),
      Map[Int, Rune](),
      Vector[Rule](),
      Map[Int, Array[Array[Int]]](),
      Map[Int, Conclusion]())
  }
}

case class SimpleSolverState[Rule, Rune, Conclusion](
  private var steps: Vector[Step[Rule, Rune, Conclusion]],

  private var userRuneToCanonicalRune: Map[Rune, Int],
  private var canonicalRuneToUserRune: Map[Int, Rune],

  private var rules: Vector[Rule],

  private var openRuleToPuzzleToRunes: Map[Int, Array[Array[Int]]],

  private var canonicalRuneToConclusion: Map[Int, Conclusion]
) extends ISolverState[Rule, Rune, Conclusion] {

  override def equals(obj: Any): Boolean = vcurious(); override def hashCode(): Int = vfail() // is mutable, should never be hashed

  def deepClone(): SimpleSolverState[Rule, Rune, Conclusion] = {
    vcurious()
    SimpleSolverState(
      steps,
      userRuneToCanonicalRune,
      canonicalRuneToUserRune,
      rules,
      openRuleToPuzzleToRunes,
      canonicalRuneToConclusion)
  }

  override def sanityCheck(): Unit = {
//    vassert(rules == rules.distinct)
  }

  override def getRule(ruleIndex: Int): Rule = rules(ruleIndex)

  override def getConclusion(rune: Rune): Option[Conclusion] = canonicalRuneToConclusion.get(getCanonicalRune(rune))

  override def getUserRune(rune: Int): Rune = canonicalRuneToUserRune(rune)

  override def getConclusions(): Stream[(Int, Conclusion)] = {
    canonicalRuneToConclusion.toStream
  }

  override def userifyConclusions(): Stream[(Rune, Conclusion)] = {
    canonicalRuneToConclusion
      .toStream
      .map({ case (canonicalRune, conclusion) => (canonicalRuneToUserRune(canonicalRune), conclusion) })
  }

  override def getAllRunes(): Set[Int] = {
    openRuleToPuzzleToRunes.values.flatten.flatten.toSet
  }

  override def addRune(rune: Rune): Int = {
    vassert(!userRuneToCanonicalRune.contains(rune))
    val newCanonicalRune = userRuneToCanonicalRune.size
    userRuneToCanonicalRune = userRuneToCanonicalRune + (rune -> newCanonicalRune)
    canonicalRuneToUserRune = canonicalRuneToUserRune + (newCanonicalRune -> rune)
    newCanonicalRune
  }

  override def getAllRules(): Vector[Rule] = {
    rules
  }

  override def addRule(rule: Rule): Int = {
    val newCanonicalRule = rules.size
    rules = rules :+ rule
//    canonicalRuleToUserRule = canonicalRuleToUserRule + (newCanonicalRule -> rule)
    newCanonicalRule
  }

  override def getCanonicalRune(rune: Rune): Int = {
    vassertSome(userRuneToCanonicalRune.get(rune))
  }

  override def addPuzzle(ruleIndex: Int, runes: Array[Int]): Unit = {
    val thisRulePuzzleToRunes = openRuleToPuzzleToRunes.getOrElse(ruleIndex, Array())
    openRuleToPuzzleToRunes = openRuleToPuzzleToRunes + (ruleIndex -> (thisRulePuzzleToRunes :+ runes))
  }

  override def getNextSolvable(): Option[Int] = {
    openRuleToPuzzleToRunes
      .filter({ case (_, puzzleToRunes) =>
        puzzleToRunes.exists(runes => {
          runes.forall(rune => canonicalRuneToConclusion.contains(rune))
        })
      })
      // Get rule with lowest ID, keep it deterministic
      .keySet
      .headOption
  }

  override def getUnsolvedRules(): Vector[Rule] = {
    openRuleToPuzzleToRunes.keySet.toVector.map(rules)
  }

  // Returns whether it's a new conclusion
  def concludeRune[ErrType](newlySolvedRune: Int, newConclusion: Conclusion):
  Result[Boolean, ISolverError[Rune, Conclusion, ErrType]] = {
    val isNew =
      canonicalRuneToConclusion.get(newlySolvedRune) match {
        case Some(existingConclusion) => {
          if (existingConclusion != newConclusion) {
            return Err(
              SolverConflict(
                canonicalRuneToUserRune(newlySolvedRune),
                existingConclusion,
                newConclusion))
          }
          false
        }
        case None => true
      }
    canonicalRuneToConclusion = canonicalRuneToConclusion + (newlySolvedRune -> newConclusion)
    Ok(isNew)
  }

  // Success returns number of new conclusions
  override def markRulesSolved[ErrType](ruleIndices: Array[Int], newConclusions: Map[Int, Conclusion]):
  Result[Int, ISolverError[Rune, Conclusion, ErrType]] = {
    val numNewConclusions =
      newConclusions.map({ case (newlySolvedRune, newConclusion) =>
        concludeRune[ErrType](newlySolvedRune, newConclusion) match {
          case Err(e) => return Err(e)
          case Ok(isNew) => isNew
        }
      }).count(_ == true)

    ruleIndices.foreach(removeRule)

    Ok(numNewConclusions)
  }

  private def removeRule(ruleIndex: Int) = {
    openRuleToPuzzleToRunes = openRuleToPuzzleToRunes - ruleIndex
  }

  class SimpleStepState(
    ruleToPuzzles: Rule => Array[Array[Rune]],
    complex: Boolean,
    rules: Vector[(Int, Rule)]
  ) extends IStepState[Rule, Rune, Conclusion] {
    private var alive = true
    private var tentativeStep: Step[Rule, Rune, Conclusion] = Step(complex, rules, Vector(), Map())

    def close(): Step[Rule, Rune, Conclusion] = {
      vassert(alive)
      alive = false
      tentativeStep
    }

    override def getConclusion(requestedUserRune: Rune): Option[Conclusion] = {
      vassert(alive)
      SimpleSolverState.this.getConclusion(requestedUserRune)
    }

    override def addRule(rule: Rule): Unit = {
      vassert(alive)
      val ruleIndex = SimpleSolverState.this.addRule(rule)
      tentativeStep = tentativeStep.copy(addedRules = tentativeStep.addedRules :+ rule)
      ruleToPuzzles(rule).foreach(puzzleUserRunes => {
        val puzzleCanonicalRunes = puzzleUserRunes.map(SimpleSolverState.this.getCanonicalRune)
        SimpleSolverState.this.addPuzzle(ruleIndex, puzzleCanonicalRunes)
      })
    }

    override def getUnsolvedRules(): Vector[Rule] = {
      vassert(alive)
      SimpleSolverState.this.getUnsolvedRules()
    }

    override def concludeRune[ErrType](newlySolvedUserRune: Rune, conclusion: Conclusion): Unit = {
      vassert(alive)
//      val newlySolvedCanonicalRune = SimpleSolverState.this.userRuneToCanonicalRune(newlySolvedUserRune)
      tentativeStep = tentativeStep.copy(conclusions = tentativeStep.conclusions + (newlySolvedUserRune -> conclusion))
//      Ok(true)
    }
  }

  override def initialStep[ErrType](
    ruleToPuzzles: Rule => Array[Array[Rune]],
    step: IStepState[Rule, Rune, Conclusion] => Result[Unit, ISolverError[Rune, Conclusion, ErrType]]):
  Result[Step[Rule, Rune, Conclusion], ISolverError[Rune, Conclusion, ErrType]] = {
    val stepState = new SimpleStepState(ruleToPuzzles, false, Vector())
    step(stepState) match {
      case Ok(()) => {
        val step = stepState.close()
        steps = steps :+ step
        Ok(step)
      }
      case Err(e) => {
        stepState.close()
        Err(e)
      }
    }
  }

  override def simpleStep[ErrType](
    ruleToPuzzles: Rule => Array[Array[Rune]],
    ruleIndex: Int,
    rule: Rule,
    step: IStepState[Rule, Rune, Conclusion] => Result[Unit, ISolverError[Rune, Conclusion, ErrType]]):
  Result[Step[Rule, Rune, Conclusion], ISolverError[Rune, Conclusion, ErrType]] = {
    val stepState = new SimpleStepState(ruleToPuzzles, false, Vector((ruleIndex, rule)))
    step(stepState) match {
      case Ok(()) => {
        val step = stepState.close()
        steps = steps :+ step
        Ok(step)
      }
      case Err(e) => {
        stepState.close()
        Err(e)
      }
    }
  }

  override def complexStep[ErrType](
    ruleToPuzzles: Rule => Array[Array[Rune]],
    step: IStepState[Rule, Rune, Conclusion] => Result[Unit, ISolverError[Rune, Conclusion, ErrType]]):
  Result[Step[Rule, Rune, Conclusion], ISolverError[Rune, Conclusion, ErrType]] = {
    val stepState = new SimpleStepState(ruleToPuzzles, true, Vector())
    step(stepState) match {
      case Ok(()) => {
        val step = stepState.close()
        steps = steps :+ step
        Ok(step)
      }
      case Err(e) => {
        stepState.close()
        Err(e)
      }
    }
  }

  override def getSteps(): Vector[Step[Rule, Rune, Conclusion]] = steps
}

File: FrontendRust/src/solver/simple_solver_state.rs
use indexmap::{IndexMap, IndexSet};
use std::collections::HashMap;
use std::collections::HashSet;
use std::hash::Hash;
use std::marker::PhantomData;

pub struct SimpleSolverState<Rule, Rune, Conclusion>
where
    Rune: Eq + Hash,
{
    rule_to_puzzles: Box<dyn Fn(&Rule) -> Vec<Vec<Rune>>>,
    steps: Vec<super::Step<Rule, Rune, Conclusion>>,
    rules: Vec<Rule>,
    all_runes: IndexSet<Rune>,
    open_rule_to_puzzle_to_runes: IndexMap<i32, Vec<Vec<Rune>>>,
    rune_to_conclusion: IndexMap<Rune, Conclusion>,
}

impl<Rule, Rune, Conclusion> SimpleSolverState<Rule, Rune, Conclusion>
where
    Rule: Clone,
    Rune: Clone + Hash + Eq,
    Conclusion: Clone + PartialEq,
{

    pub fn sanity_check(&self) {
        // vassert(rules == rules.distinct)
    }

    pub fn get_rule(&self, rule_index: i32) -> &Rule {
        &self.rules[rule_index as usize]
    }

    pub fn get_conclusion(&self, rune: &Rune) -> Option<Conclusion> {
        self.rune_to_conclusion.get(rune).cloned()
    }

    pub fn get_conclusions(&self) -> Vec<(Rune, Conclusion)> {
        self.rune_to_conclusion.iter().map(|(k, v)| (k.clone(), v.clone())).collect()
    }

    pub fn userify_conclusions(&self) -> Vec<(Rune, Conclusion)> {
        self.rune_to_conclusion.iter().map(|(k, v)| (k.clone(), v.clone())).collect()
    }

    pub fn get_all_runes(&self) -> HashSet<Rune> {
        self.all_runes.iter().cloned().collect()
    }

    pub fn is_complete(&self) -> bool {
        self.rune_to_conclusion.len() == self.all_runes.len()
    }

    pub fn commit_step<ErrType>(
        &mut self,
        complex: bool,
        solved_rule_indices: Vec<i32>,
        conclusions: IndexMap<Rune, Conclusion>,
        new_rules: Vec<Rule>,
        new_runes: IndexSet<Rune>,
    ) -> Result<(), super::ISolverError<Rune, Conclusion, ErrType>> {
        self.all_runes.extend(new_runes);
        let solved_rules: Vec<(i32, Rule)> = solved_rule_indices
            .iter()
            .map(|&idx| (idx, self.rules[idx as usize].clone()))
            .collect();
        let step = super::Step {
            complex,
            solved_rules,
            added_rules: new_rules.clone(),
            conclusions: conclusions.clone().into_iter().collect(),
        };
        // Append step before checking for conflicts (audit trail captures conflicting step)
        self.steps.push(step);
        // Check and apply conclusions
        for (rune, new_conclusion) in &conclusions {
            if let Some(existing) = self.rune_to_conclusion.get(rune) {
                if existing != new_conclusion {
                    return Err(super::ISolverError::SolverConflict(
                        super::SolverConflict {
                            rune: rune.clone(),
                            previous_conclusion: existing.clone(),
                            new_conclusion: new_conclusion.clone(),
                            _phantom: PhantomData,
                        },
                    ));
                }
            }
            self.rune_to_conclusion.insert(rune.clone(), new_conclusion.clone());
        }
        // Remove solved rules
        for &rule_index in &solved_rule_indices {
            self.open_rule_to_puzzle_to_runes.remove(&rule_index);
        }
        // Add new rules with puzzles
        for rule in new_rules {
            let rule_index = self.rules.len() as i32;
            self.rules.push(rule.clone());
            let puzzles = (self.rule_to_puzzles)(&rule);
            for puzzle in &puzzles {
                for rune in puzzle {
                    assert!(self.all_runes.contains(rune), "vassert: rune in puzzle must be in allRunes");
                }
            }
            self.sanity_check();
            for puzzle in puzzles {
                let entry = self.open_rule_to_puzzle_to_runes
                    .entry(rule_index)
                    .or_insert_with(Vec::new);
                entry.push(puzzle);
            }
            self.sanity_check();
        }
        Ok(())
    }

    pub fn get_next_solvable(&self) -> Option<i32> {
        // Get rule with lowest ID, keep it deterministic (matches Scala)
        self.open_rule_to_puzzle_to_runes
            .iter()
            .filter(|(_, puzzle_to_runes)| {
                puzzle_to_runes.iter().any(|runes| {
                    runes
                        .iter()
                        .all(|r| self.rune_to_conclusion.contains_key(r))
                })
            })
            .map(|(rule_index, _)| *rule_index)
            .min()
    }

    pub fn get_unsolved_rules(&self) -> Vec<Rule> {
        self.open_rule_to_puzzle_to_runes
            .keys()
            .map(|&idx| self.rules[idx as usize].clone())
            .collect()
    }

    pub fn get_unsolved_runes(&self) -> Vec<Rune> {
        self.all_runes.iter()
            .filter(|r| !self.rune_to_conclusion.contains_key(*r))
            .cloned()
            .collect()
    }

    pub fn get_steps(&self) -> Vec<super::Step<Rule, Rune, Conclusion>> {
        self.steps.clone()
    }

    pub fn rule_is_solved(&self, rule_index: i32) -> bool {
        !self.open_rule_to_puzzle_to_runes.contains_key(&rule_index)
    }

pub fn new(
    rule_to_puzzles: Box<dyn Fn(&Rule) -> Vec<Vec<Rune>>>,
    all_runes: Vec<Rune>,
) -> Self {
    SimpleSolverState {
        rule_to_puzzles,
        steps: vec![],
        rules: vec![],
        all_runes: all_runes.into_iter().collect(),
        open_rule_to_puzzle_to_runes: IndexMap::new(),
        rune_to_conclusion: IndexMap::new(),
    }
}

}
